code/datums/climate.dm

```diff
@@ -0,0 +1,209 @@
+
+/datum/climate/proc/tick()
+	if(!current_weather)
+		return
+	current_weather.tick()
+	if(current_weather.timeleft <= 0)
+		change_weather(forecasts[1])
+		forecasts -= forecasts[1]
+	if(forecasts.len < PREDICTION_MINIMUM)
+		forecast()
+
+/datum/climate/proc/change_weather(weather)
+	if(ispath(weather))
+		//We have been provided a path. Let's see if it's identical to the one we have.
+		if(ispath(weather, current_weather.type)) //This is a separate check so that we can have our warning work.
+			return //No need to change, this is our current type.
+		else
+			qdel(current_weather)
+			current_weather = new weather(src)
+			current_weather.execute()
+
+	else if(istype(weather,/datum/weather))
+		//We have been given a specific weather datum. It may be modified, so run it no matter what.
+		qdel(current_weather)
+		current_weather = weather
+		current_weather.execute()
+
+	else
+		WARNING("Change weather was called with [weather], neither a weather datum nor a path.")
+
+/datum/climate/snow
+	//some day this may not be the norm
+
+/datum/climate/snow/New()
+	current_weather = new /datum/weather/snow/calm(src)
+	..()
+
+///////////////////////////////////  WEATHER DATUMS //////////////////////////////
+
+/datum/weather
+	var/name = "weather"
+	var/list/next_weather = list() //associative list
+	var/timeleft = 1
+	var/datum/climate/parent
+	var/temperature = T20C
+
+/datum/weather/New(var/datum/climate/C)
+	parent = C
+	timeleft = round(rand(parent.cycle_freq[1],parent.cycle_freq[2]),SS_WAIT_WEATHER)
+	//round to 2 seconds, since that's how often we check in
+
+/datum/weather/proc/execute()
+
+/datum/weather/proc/tick()
+	timeleft -= SS_WAIT_WEATHER
+
+var/list/global_snowtiles = list()
+var/list/snow_state_to_texture = list()
+var/snowtiles_setup = 0 //has the blizzard parent been initialized?
+
+/datum/weather/snow
+	var/snow_intensity = SNOW_CALM
+	var/tile_interval = 5
+	var/snowfall_prob = 0
+	var/snowfall_rate = list(0,0)
+	var/snow_fluff_estimate = "snowing"
+
+/datum/weather/snow/execute()
+	for(var/obj/machinery/teleport/hub/emergency/E in machines)
+		E.alarm(!(snow_intensity % SNOW_BLIZZARD))
+		//sends 1 if snow_intensity equals blizzard exactly, otherwise sends 0
+	for(var/turf/unsimulated/floor/snow/tile in global_snowtiles)
+		if(tile.ignore_blizzard_updates)
+			continue
+		tile.snow_state = snow_intensity
+		tile.update_environment()
+	force_update_snowfall_sfx()
+
+/datum/weather/snow/tick()
+	..()
+	if(!prob(snowfall_prob))
+		return
+	var/i = rand(1,tile_interval)
+	for(var/turf/unsimulated/floor/snow/tile in global_snowtiles)
+		if(i == tile_interval)
+			tile.change_snowballs(snowfall_rate[1],snowfall_rate[2])
+			if(tile.snowprint_parent)
+				tile.snowprint_parent.ClearSnowprints()
+			i = 1
+		else
+			i++
+
+var/list/snowstorm_ambience = list('sound/misc/snowstorm/snowfall_calm.ogg','sound/misc/snowstorm/snowfall_average.ogg','sound/misc/snowstorm/snowfall_hard.ogg','sound/misc/snowstorm/snowfall_blizzard.ogg')
```

Comment:
Does this changeset fix the issue where you would hear snow even after even after it ended, or even after you moved into another area?

Reply:
no, I think the only way to really fix that is going to be to break it into smaller, more manageable sounds and that's for another day and someone who has audacity installed

Reply:
If you assign the sound to a specific channel, you can stop it:

```dm
// I think BYOND has a channel limit but can't recall what it is.
#define CHANNEL_WEATHER 10 // Snow, rain, etc
// Snow sounds
usr << sound('sound/weather/snow_average.ogg', channel=CHANNEL_WEATHER)
// Stop snow sounds
usr << sound(null, channel=CHANNEL_WEATHER)

Reply:
yes, it's already on the weather channel, unfortunately as far as I can tell, nowhere does it call null and worse it only plays the (long) soundfile to you if you happened to be on snow when it updated

basically it's glasscode and needs to be rooted up in its own PR with a solid rewrite